olivertwist/rules/max_two_predecessor_views.py

```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+"""
+Table/incremental materialized models should have a maximum of two views as predecessors before
+being materialized as tables. If this is not enforced Spark can struggle with creating query
+plans for long lineage chains of views.
+
+This test is a little more complex than normal as it involves tracing the lineage from a model
+through multiple node layers whereas normally we only look back at a single predecessor or
+successor. Using networkx lineage we need to look at preceding nodes up to 3 layers back.
+If we have a consistent chain of 3 views at this stage we can fail the test.
+"""
+
+from typing import List, Tuple
+
+from olivertwist.manifest import Manifest, Node
+from olivertwist.ruleengine.rule import rule
+from olivertwist.rules.utils import partition
+
+
+@rule(
+    id="max-two-predecessor-views",
+    name="Models can only have two or less predecessor views in a chain",
+)
+def models_have_more_than_two_predecessor_views(
+    manifest: Manifest,
+    max_view_depth=2
+) -> Tuple[List[Node], List[Node]]:
+    def model_has_more_than_two_predecessor_views(
+        node: Node
+    ) -> bool:
+        def get_predecessor_layer_view_count(layer_nodes, layer_depth):
+            predecessor_layer_nodes = []
+            for node in layer_nodes:
+                for node_predecessor_node in manifest.graph.predecessors(node):
+                    if manifest.get_node(node_predecessor_node).materialization in ['view','ephemeral']:
+                        predecessor_layer_nodes.append(node_predecessor_node)
+                        layer_depth = max(layer_depth, layer_depth+1)
+            return predecessor_layer_nodes, layer_depth
+
+        current_layer_nodes = [node.id]
+        max_layer_depth = 0
+        for i in range(max_view_depth+1):
+            current_layer_nodes, max_layer_depth = get_predecessor_layer_view_count(current_layer_nodes, max_layer_depth)
+            if max_layer_depth > max_view_depth:
+                return True
+        return False
```

Comment:
I'm a bit uneasy having two nested closures here, I think it really detriments readability unnecessarily.  What would you think about moving these functions to their own scope i.e.
```
def get_predecessor_layer_view_count(manifest, layer_nodes, layer_depth):
            predecessor_layer_nodes = []
            for node in layer_nodes:
                for node_predecessor_node in manifest.graph.predecessors(node):
                    if manifest.get_node(node_predecessor_node).materialization in ['view','ephemeral']:
                        predecessor_layer_nodes.append(node_predecessor_node)
                        layer_depth = max(layer_depth, layer_depth+1)
            return predecessor_layer_nodes, layer_depth
```
you could do the same with `model_has_more_than_two_predecessor_views(manifest, max_view_depth, node)` and use `functools.partial` to still get a functional approach while improving readability a bit